File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/statistics/StatisticsPresenter.java
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.statistics;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.zhuinden.simplestackdemoexamplefragments.data.repository.TaskRepository;
import com.zhuinden.simplestackdemoexamplefragments.presentation.objects.Task;
import com.zhuinden.simplestackdemoexamplefragments.util.BasePresenter;
import com.zhuinden.statebundle.StateBundle;

import org.javatuples.Pair;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Owner on 2017. 01. 27..
 */

public class StatisticsPresenter
        extends BasePresenter<StatisticsFragment, StatisticsPresenter> {
    private final TaskRepository taskRepository;

    @Inject
    public StatisticsPresenter(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    Disposable disposable;

    @Override
    protected void onAttach(StatisticsFragment coordinator) {
        disposable = Observable.combineLatest(taskRepository.getActiveTasksWithChanges(),//
                                              taskRepository.getCompletedTasksWithChanges(),//
                                              (activeTasks, completedTasks) -> Pair.with(activeTasks, completedTasks)) //
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(pairOfActiveAndCompletedTasks -> {
                    List<Task> activeTasks = pairOfActiveAndCompletedTasks.getValue0();
                    List<Task> completedTasks = pairOfActiveAndCompletedTasks.getValue1();
                    if(getFragment() != null) {
                        getFragment().showStatistics(activeTasks.size(), completedTasks.size());
                    }
                });
    }

    @Override
    protected void onDetach(StatisticsFragment coordinator) {
        disposable.dispose();
    }

    @Override
    @NonNull
    public StateBundle toBundle() {
        return new StateBundle();
    }

    @Override
    public void fromBundle(@Nullable StateBundle bundle) {

    }
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/statistics/StatisticsPresenter.kt
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.statistics

import com.zhuinden.simplestackdemoexamplefragments.data.repository.TaskRepository
import com.zhuinden.simplestackdemoexamplefragments.util.BasePresenter
import com.zhuinden.simplestackdemoexamplefragments.util.BaseViewContract
import com.zhuinden.simplestackdemoexamplefragments.util.combineTwo
import com.zhuinden.statebundle.StateBundle
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

class StatisticsPresenter @Inject constructor(
    private val taskRepository: TaskRepository
) : BasePresenter<StatisticsPresenter.ViewContract>() {
    interface ViewContract: BaseViewContract {
        fun showStatistics(numberOfIncompleteTasks: Int, numberOfCompletedTasks: Int)
    }

    private lateinit var disposable: Disposable

    override fun onAttach(view: StatisticsPresenter.ViewContract) {
        disposable = combineTwo(
            taskRepository.activeTasksWithChanges,
            taskRepository.completedTasksWithChanges)
            .subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { (activeTasks, completedTasks) ->
                view.showStatistics(activeTasks.size, completedTasks.size)
            }
    }

    override fun onDetach(view: StatisticsPresenter.ViewContract) {
        disposable.dispose()
    }

    override fun toBundle(): StateBundle = StateBundle()

    override fun fromBundle(bundle: StateBundle?) {
    }
}
